Agent Installer Development Environment (#1034)
openshift/enhancements: enhancements/agent-installer/development-environment.md

---

> title: ephemeral-agent-installer-e2e-tests

Reviewer: ```suggestion
title: development-environment
```

- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Create a development environment for the ephemeral agent based installer image in the 
[dev-scripts](https://github.com/openshift-metal3/dev-scripts) repository and refactor 
components as needed for the new project. In addition we will add base case e2e test flows
integrated into Openshift CI to get the project started.

## Motivation

The new agent based installer project will need a standard development environment for
the team to use. We will need an e2e testing framework for validation in Openshift CI. 
In this enhancement we will explain our desire to reuse existing frameworks for the new 
project while limiting the scope of modifications to them to the new project's requirements.

### Goals

* Create a development environment for the agent installer project that can be used for local
development and for automating the flows in Openshift CI.
* Create predefined configurations to deploy Openshift in HA, compact and SNO topologies.
* Create predefined network configurations for Openshift including but not limited IP type 

> (IPv4, IPv6, dual-stack), L2 segments and DHCP configurations.

Reviewer: Will we say the enhancement isn't fully implemented until all of these profiles/configurations are available? Or is the goal to enable them to be created over time?

Author: My view is that we absolutely need this set because this is most the common configurations we use for OCP installs. The two that aren't: L2 segments and DHCP are linked in with work that is planned for Agent Installer. So it is fair to say we aren't complete until we get these in.

Reviewer: Yeah, I'm not questioning whether need them. I'm questioning whether we would say "we have neither a development environment tool nor CI" until those are written.

It's safe to break this work down into more steps, and for this enhancement to say the goal is to provide 1 basic configuration and then enable us to build those "advanced" configurations later. You wouldn't need separate enhancements for the later work, but you might track them in separate tickets.

Author: Ahh okay I understand the question now. 🤔 I'll split them into different phases. It isn't too hard to do most of them because a lot of this already defined in Openshift CI

Author: Yeah that's a good point with disconnected vs connected the changes between the two are pretty advanced configurations 🤔

> * A reusable Ansible environment configuration for development and testing teams to use with

Reviewer: ```suggestion
* A reusable environment configuration for development and testing teams to use with
```

Let's not pre-select the implementation tool in the goal.

predefined flows for quicker onboarding.
* Convergence of the underlying scripts and environment setup code between Metal Platform (IPI) 
team and the new Agent Installer teams.

### Non-Goals

* Re-factoring dev-scripts and assisted-test-infra for reasons not related to the new 
ephemeral agent based installer image project.
* Overall convergence between all the installer development teams of scripts and e2e test flows.

## Proposal

Create an development environment for the ephemeral agent based installer image in the 
[dev-scripts](https://github.com/openshift-metal3/dev-scripts) repository and refactor 
components as needed for the new project.

### User Stories

> * As a developer, I need to set up an environment so that I can simulate the installation of a 
> users first cluster using the ephemeral based agent installer image.

Reviewer: ```suggestion
* As a developer, I need to set up an environment so that I can install a 
cluster locally using the agent-based installer for development and testing
```

* As a developer, I need to launch the built ephemeral agent based installer image so that I can 
test and validate this image during development.

* As a developer, I need to check cluster installation progress of the Openshift cluster being
installed by agent installer and when this installation is complete to validate the Openshift
cluster.

* As a developer, I want to extract logs from agent installer and the Openshift cluster under

> install to troubleshoot and debug.

Reviewer: Will this goal influence the design of the solution?

Author: It shouldn't influence the design of the development environment. The development environment would just be using existing API calls against assisted-service, must-gather against the Openshift cluster, etc to get this information.

Reviewer: OK. You can leave it in, but generally we want to limit the goals list to those that apply directly to the planned work so we can map the proposed changes back to the goals and make sure we're meeting them all.

### API Extensions

Not applicable.

### Risks and Mitigations

** There will be now a third team using dev-scripts for slightly different purposes.
Which means the risk of breaking things in the central framework now is larger and
has a larger "splash zone."

> We can mitigate this for the new ephemeral agent installer by adding e2e tests into the
> dev-scripts repository that run the agent flows. The purpose of this would be to find 

Reviewer: ```suggestion
We can mitigate this for the new ephemeral agent installer by running e2e tests using the agent flows for changes to the
dev-scripts repository. The purpose of this would be to find 
```

breakages in the tests across the different teams.

** The dev-scripts repository clones the [metal-dev-env](https://github.com/metal3-io/metal3-dev-env) 
repository for most of its Ansible roles. Which means there is a risk of changes in this 
repository can break things in our repository as well. We have less control over this repository 
than others inside Openshift.

This is partially mitigated already because the repository is cloned at a specific
Git SHA. To further this we can add a test in dev-scripts that runs maybe once or twice
a week that clones the HEAD of this repository. We would be alerted for breakages more
frequently and we can mitigate them on a case-by-case basis.

## Design Details

> ### General Strategy

Reviewer: I like the phase breakdown.

* Using the scope of the agent installer project we will identify possible refactors to 
existing scripts and new functionality required by this new project.
* New functionality we will write in Ansible, unless it is highly coupled with existing
shell scripts.

> * We will create Anisble roles to wrap the existing code making it easier to use within

Reviewer 2: typo: Ansible

a pure Ansible solution
* The existing bash scripts will be left as is. Refactoring them will be done piece by 
piece as needed.

### Phase 1 (Crawl): Add development scripts into dev-scripts for Agent Installer

> 1) Add development scripts into dev-scripts for the new agent installer project using 

Reviewer 3: Can you clarify what are the development scripts? Perhaps explain their purpose.

Author: I added the three scenarios in an update. This is not meant to be an exhaustive list just to get us started.

existing conventions.
2) Add Ansible roles and tasks to wrap existing shell scripts to make them easier
to integrate with an Ansible solution.
3) Add base e2e test flow to run the ephemeral agent installer image in Openshift CI.

> 4) Write documentation for new developers joining the agent installer team.

Reviewer: Documentation in dev-scripts?

Author: Yes. It just means add new documentation for the new things we are adding for agent installer. How to use dev-scripts to setup your development environment to write/test agent installer. If we don't want to call it out I can remove it.

Reviewer: It's a good thing to call out, I was just curious where those docs would go.

In this phase it should also become more clear what scripts and code are overlapping

> between the two teams.

Reviewer 4: nit: more than "between the two teams" I'd suggest to change it to "with the existing code base" (ie right now more than two teams are using dev-scripts for their dev / test environments)

and networks (IPV4, IPV6, dualstack) in Ansible.
4) Identify minor refactors to existing scripts deemed helpful for the two projects and
move them into Ansible.

### Phase 3 (Run): Update existing e2e IPI flows to use the refactored Ansible in dev-scripts

1) Update Agent Installer flows to use the new Ansible roles to configure the flows instead
of using the environment variables directly.
2) Update the e2e IPI flows to use this new Ansible configuration roles as needed.

### Open Questions [optional]

### Test Plan

* Test locally while creating the development environment.
* On-board the base e2e flow into Openshift CI for agent installer.

### Graduation Criteria

Not applicable.

#### Removing a deprecated feature

Not applicable.

### Upgrade / Downgrade Strategy

Not applicable.

### Version Skew Strategy

* Currently the dev-scripts repository is not versioned. This will stay the same.

* However for the e2e test scenarios themselves it makes sense to version these.
Since the ephemeral agent based installer is targeted to be released with OCP it would 
be easiest to version the e2e framework the same as OCP x.y. This is the strategy

> that openshift-tests takes.

Reviewer: Do you expect to need new tests? Or a new test job that uses the agent-based installer to set up the cluster and then run the existing e2e tests?

Author: For this enhancement the strategy I'm suggesting is the latter. If we find we need new test scenarios we can revise the enhancement to include them. Reusing what we have already is more of a baseline we need to get to before adding new tests.

Author: I moved this information into the design details section under phase 1 in the update.

project while limiting the scope of modifications to them to the new project's requirements.

In this space we have two existing frameworks that could be used for this new project. These 
frameworks are:
* [dev-scripts](https://github.com/openshift-metal3/dev-scripts): Created by the Metal Platform 
(IPI) team to set up environments to install OCP both for local development and in Openshift CI.
* [assisted-test-infra](https://github.com/openshift/assisted-test-infra): Created by the Assisted
Installer team to run full e2e test flows.

Looking at existing e2e test flow steps from these two frameworks it was determined that 
in dev-scripts test flows the beginning steps are very similar to the development
flows we desire for the agent installer. In contrast the assisted-test-infra test flows
set up extra things in the development environment that is undesirable for the agent installer.

### Goals

* Using this development environment in the new ephemeral agent based installer image project.
* Re-factoring parts of dev-scripts that are needed for this ephemeral agent based installer.
* Convergence of the underlying scripts and environment setup code between Metal Platform (IPI) 

> team and the new Agent Installer teams.

Reviewer: I think this thread can be resolved?

### Non-Goals

* Re-factoring dev-scripts and assisted-test-infra for reasons not related to the new 
ephemeral agent based installer image project.
* Overall convergence between all the installer development teams of scripts and e2e test flows.

## Proposal

Create an development environment for the ephemeral agent based installer image in the 
[dev-scripts](https://github.com/openshift-metal3/dev-scripts) repository and refactor 
components as needed for the new project.

### User Stories

> * As a developer, I need to set up an environment so that I can simulate the installation of a 

Reviewer 5: I'd say something like "emulate on-premise customer environments (including various network and server configurations), to test installation using the ephemeral agent based installer"

> * As a developer, I need to set up an environment so that I can simulate the installation of a 

Author: I'll put in the adjective change here. :) 

But, just to give some context the use of simulate vs emulate is probably because of my hardware background. Hardware emulation vs Software simulation is the context I'm used to having these words. In that context these are pretty different things despite both being software programs. What we are doing with the VM environment for installation is close to "a software based simulation of a customers installation environment." That's how I would describe it... but that isn't too far off from Steve's wording either.

> * As a developer, I need to set up an environment so that I can simulate the installation of a 

Reviewer 2: It's a real installation into a simulated environment (VMs and linux bridges instead of servers and switches).

users first cluster using the ephemeral based agent installer image.

* As a developer, I need to launch the built ephemeral agent based installer image so that I can test 
and validate this image during development.

* As a developer, I need to install Openshift using the ephemeral agent based installer image 

> to validate the agent installer.

Author: Yeah I cut out some of them since it seemed redundant.

* As a developer, I need to check cluster installation progress of the Openshift cluster being
installed by agent installer and when this installation is complete to validate the Openshift

> cluster.

Author: It's not really included in how you write a user story but I plan to use the existing Openshift tests to validate the installed cluster.

> cluster.

Reviewer 4: I think that (in the long term) the number of configurations * OpenShift versions * e2e conformance suits required for an exhaustive vaidation will make it impractical to verify it locally, leaving this responsibility to the CI rehearsals.
During the development activity I'd like to have a quick / gross way to verify the current changes, so probably we could identify a number of minimum must-have checks required - and possibly common to all the possible configurations (ie cluster is up), but that's probably something we could leave as an implementation details and here just leave "verify that installation is complete"

> cluster.

Author: I like this idea of defined minimum checks. That would help with onboarding new developers and learning to develop against OCP. This could be something we define for the future... We should at least put it in the backlog.

My desire to have the configurations local is more about running the configuration I'm currently working on while I'm developing. I want to have that ability to do that on demand and easily. For instance if I'm working on a feature that is specific to SNO clusters, I would only run SNO based flows locally while developing. When I push up my work the CI will run compact cluster and HA clusters checks against my code to make sure I didn't break anything. After that is finished I might work on a feature that is focused on a different type of cluster configuration than SNO so I wouldn't run SNO based flows. I realize this is my personal preference in developing but I think we can accommodate both styles without much issue.

> cluster.

Author: I also just realized this thread is more about validations of the OCP clusters than configurations.

Local validations I'd like to see:
- Kubeconfig is created and accessible
- Control Plane came up
- Cluster is installed
- A generic way to capture the error output when cluster fails to install

* As a developer, I need to validate the kubeconfig is available so I can use it to access the
installed cluster.

* As a developer, I want to extract logs from agent installer and the Openshift cluster under
install to troubleshoot and debug.

### API Extensions

Not applicable.

### Risks and Mitigations

The direction of this proposal is now to extend and refactor parts of dev-scripts
for the agent-based installer.

** There will be now a third team using dev-scripts for slightly different purposes.
Which means the risk of breaking things in the central framework now is larger and
has a larger "splash zone."

We can mitigate this for the new ephemeral agent installer by adding e2e tests into the

> dev-scripts repository that run the agent flows. The purpose of this would be to find 

Reviewer 5: I think it means add e2e tests that run on PRs proposed to the dev-scripts repository

> dev-scripts repository that run the agent flows. The purpose of this would be to find 

Author: Yes that's what I mean. Just put the e2e tests that use dev-scripts as a base on the dev-scripts repo so that PRs into dev-scripts don't break them.

breakages in the tests across the different teams.

** The dev-scripts repository clones the [metal-dev-env](https://github.com/metal3-io/metal3-dev-env) 
repository for most of its Ansible roles. Which means there is a risk of changes in this 
repository can break things in our repository as well. We have less control over this repository 
than others inside Openshift.

This is partially mitigated already because the repository is cloned at a specific
Git SHA. To further this we can add a test in dev-scripts that runs maybe once or twice
a week that clones the HEAD of this repository. We would be alerted for breakages more

> frequently and we can mitigate them on a case-by-case basis.

Reviewer 5: So far we haven't found this additional testing to be necessary, but we do sometimes have to make changes to dev-scripts when moving the metal3-dev-env pin (for example increasing ansible/python version)

Reviewer: Let's not sign up for more work than we need. I like the idea of keeping this as a suggestion, but not a commitment of something we'll do.

Author: Okay I'll pair this one down to be a suggestion.

## Design Details

### Phase 1 (Crawl): Add development scripts into dev-scripts for Agent Installer

1) Add development scripts into dev-scripts for the new agent installer project using 
existing conventions.

> 2) Add base e2e test flow to run the ephemeral agent installer image in Openshift CI.

Author: I left this open because we are still deciding where things go. Right now dev-scripts and fleeting for sure. Looks like we will soon add another to this list now. ;)

3) Write documentation for new developers joining the agent installer team.

In this phase it should also become more clear what scripts and code are overlapping
between the two teams.

### Phase 2 (Walk): Update shell scripts to Ansible in dev-scripts and use them for Agent Installer

1) Convert the shell scripts [01_install_requirements.sh](https://github.com/openshift-metal3/dev-scripts/blob/master/01_install_requirements.sh) and [02_configure_host.sh](https://github.com/openshift-metal3/dev-scripts/blob/master/02_configure_host.sh) 

> into Ansible roles and playbooks for reuse.

Reviewer 5: Yeah this seems desirable, but unrelated to the agent integration

> into Ansible roles and playbooks for reuse.

Author: This was cut in the rewording of this section :)

common configurations for OCP installs. 
3) Use the above examples to define default configurations for VMs (HA, Compact, SNO) 
and networks (IPV4, IPV6, dualstack) in Ansible.

### Phase 3 (Run): Update existing e2e IPI flows to use the refactored Ansible in dev-scripts

Update the existing e2e IPI flows to use the new Ansible roles and playbooks and debug
to ensure they work as expected. Doing this in a separate phase will allow us to explore
new options for the Agent Installer project while maintaining healthy e2e flows for IPI.
This should allow for more collaboration between the two teams.

### Refactored File and Folder Structure

The following is describing general pattern of how to organize these things in the dev-scripts 
repository. These files and folders will be added alongside existing scripts in the dev-scripts 
repository. This example includes things that have been already identified as overlaps between 
the two teams that will need to be added or refactoring. Things may change names or locations
during the code review process this folder structure is added for clarification.

> **File and Folder Structure:**

Reviewer 4: I think that here there could be the risk to over-engineering the solution, considering that in future things may evolve differently (for example, integration within the openshift installer). In addition there are parts that seem to overlap with already existing frameworks such as the `ci-operator` and already defined steps at OpenShift level (ie ci-operator steps registry)

Reviewer 2: We'll need a separate step to https://github.com/openshift/release/blob/master/ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-commands.sh (just running `make` won't cut it), so what's the harm in putting most of the configuration for each topology/network scenario into the dev-scripts repo where developers can also make use of it, and having just 1 configuration option (i.e. which scenario to choose) in the ci-operator steps?

Author: This is the route I went with on the rewording of it. Closer to the intention of it all as well.
